perf: Add HTTP Link header for preload hints

adityahase opened the pull request:
Parse rendered templates to find stylesheet and script files, add these files to the HTTP Link header.

In the case of HTTP 1.1 browser can fetch these resources before parsing the DOM.
In the case of HTTP 2.0 NGINX can use HTTP2 Server Push (`http2_push_preload on;`).

#### Before
**HTTP 1.1 without preload hints**
![http1 1-nopreload](https://user-images.githubusercontent.com/8528887/81038708-a3bc9200-8ec4-11ea-88f7-f007f14f00a3.png)

#### After
**HTTP 1.1 with preload hints**
![http1 1-preload](https://user-images.githubusercontent.com/8528887/81038714-a9b27300-8ec4-11ea-8b92-06cd5e8c56a1.png)

**HTTP 2 with preload hints with Server Push**
![http2-preload-push](https://user-images.githubusercontent.com/8528887/81038754-c9e23200-8ec4-11ea-9fd5-b1e985094558.png)


**Note:** Use `listen ssl http2;` and`http2_push_preload on;` to enable Server Push on NGINX.

**Note:** **Finish**, **DomContentLoaded** and **Load** times show significant variation on repeated runs. To replicate, use production setup.

Reference: https://www.w3.org/TR/preload/

netchampfaris commented on frappe/utils/jinja.py:
```diff
@@ -77,15 +77,40 @@ def render_template(template, context, is_path=None, safe_render=True):
 		except TemplateError:
 			throw(title="Jinja Template Error", msg="<pre>{template}</pre><pre>{tb}</pre>".format(template=template, tb=get_traceback()))
 
+def add_preload_hints(content):
```
You can do this in `build_page` after all of the HTML has been rendered:

https://github.com/frappe/frappe/blob/9c7f01b59d69799eaae5354f49c44d67c87f6091/frappe/website/render.py#L184

netchampfaris commented on frappe/website/render.py:
```diff
@@ -86,7 +86,9 @@ def render(path=None, http_status_code=None):
 		})
 
-	return build_response(path, data, http_status_code or 200)
+	headers = {"Link": ",".join(frappe.local.preload)} if hasattr(frappe.local, "preload") else {}
```
Shouldn't we preload at the HTML content level itself? Preload is used to prioritise certain resources over others. If all of them are prioritised do we gain anything?

https://developer.mozilla.org/en-US/docs/Web/HTML/Preloading_content

adityahase replied:
What I'm trying to achieve here is HTTP 2 Server Push, It can't function with NGINX without the Link header.

adityahase committed:
perf: Add HTTP Link header for preload hints

Parse rendered templates and find stylesheet and script files.
Add these files to the HTTP Link header.

In the case of http1.1 browser can fetch these resources before parsing the DOM.
In the case of http2 NGINX can use HTTP2 Server Push (http2_push_preload on).

https://www.w3.org/TR/preload/

adityahase commented:
@Mergifyio backport version-12-hotfix
